Improve rake evm:status_full

kbrock opened the pull request:
Currently, it is difficult to get a quick overview of a customer's appliance setup.
This introduces a way to quickly describe a customer's appliance setup.

It is grouped by region/zone, and similar to `evm:server_status`.

The 2 tables display versions and worker counts to quickly understand the layout more than the actual health.

1. Start and heartbeat have been shortened. It shows the time if it is less than 24 hours ago. If it started longer, then it displays the date. The premise is if it has been running for a few days, the exact time it started is not significant. Knowing if it recently started may be indicative of a memory threshold issue.
2. If all rows have the same value in a column, the column is put into a summary below. This helps reduce the number of columns on the screen and hopefully avoid the line wrap.
3. Version has been added to know of of outstanding upgrades.
4. Number of workers added
5. ~~Status has been simplified to up/down and ? (unresponsive).~~
5. For this example, they all have the same Region, Version, and Started On.

### before

```
Checking EVM status...
 Zone           | Server             | Status         | ID |   PID |  SPID | URL                     | Started On           | Last Heartbeat       | MB Usage | Master? | Active Roles
----------------+--------------------+----------------+----+-------+-------+-------------------------+----------------------+----------------------+----------+---------+---------------------------------------
 default        | cfme-5.9.0.17-TE12 | started        | 11 |  2628 | 14023 | druby://127.0.0.1:43395 | 2018-01-18T23:27:43Z | 2018-01-21T01:29:45Z |      532 | false   | user_interface:web_services:websocket
 default        | cfme-5.9.0.17-TE1  | started        |  1 | 12783 | 12909 | druby://127.0.0.1:35522 | 2018-01-18T23:11:24Z | 2018-01-21T01:30:02Z |      251 | true    | user_interface:web_services:websocket
 default        | cfme-5.9.0.17-TE11 | started        | 12 | 13449 | 13622 | druby://127.0.0.1:34063 | 2018-01-18T23:26:42Z | 2018-01-21T01:29:52Z |      605 | false   | user_interface:web_services:websocket
 Azure Zone     | cfme-5.9.0.17-TE9  | started        |  9 |   415 | 13591 | druby://127.0.0.1:34285 | 2018-01-18T23:26:29Z | 2018-01-21T01:29:52Z |      298 | false   | user_interface:web_services:websocket
 Amazon Zone    | cfme-5.9.0.17-TE6  | started        |  5 | 20323 | 13488 | druby://127.0.0.1:40147 | 2018-01-18T23:25:18Z | 2018-01-21T01:29:27Z |      259 | false   | user_interface:web_services:websocket
 Config Zone    | cfme-5.9.0.17-TE10 | started        | 10 | 13064 | 13604 | druby://127.0.0.1:42435 | 2018-01-18T23:26:35Z | 2018-01-21T01:30:01Z |      277 | false   | user_interface:web_services:websocket
 OpenShift Zone | cfme-5.9.0.17-TE3  | started        |  4 | 15253 | 13516 | druby://127.0.0.1:36287 | 2018-01-18T23:25:13Z | 2018-01-21T01:28:48Z |      265 | false   | user_interface:web_services:websocket
 OpenStack Zone | cfme-5.9.0.17-TE2  | started        |  2 | 14034 | 13507 | druby://127.0.0.1:45789 | 2018-01-18T23:25:58Z | 2018-01-21T01:29:41Z |      515 | false   | user_interface:web_services:websocket
 RHV Zone       | cfme-5.9.0.17-TE4  | not responding |  7 | 31092 | 13544 | druby://127.0.0.1:39409 | 2018-01-18T23:26:14Z | 2018-01-19T11:53:35Z |      237 | false   |
```

### after
```
Checking EVM status...
 Zone           | Server             | Status         |   PID |  SPID | Workers | Heartbeat  | MB Usage | Roles
----------------+--------------------+----------------+-------+-------+---------+------------+----------+---------------------------------------
 Amazon Zone    | cfme-5.9.0.17-TE6  | started        | 20323 | 13488 |      19 | 2018-01-21 |      259 | user_interface:web_services:websocket
 Azure Zone     | cfme-5.9.0.17-TE9  | started        |   415 | 13591 |      17 | 2018-01-21 |      298 | user_interface:web_services:websocket
 Config Zone    | cfme-5.9.0.17-TE10 | started        | 13064 | 13604 |      18 | 2018-01-21 |      277 | user_interface:web_services:websocket
 OpenShift Zone | cfme-5.9.0.17-TE3  | started        | 15253 | 13516 |      18 | 2018-01-21 |      265 | user_interface:web_services:websocket
 OpenStack Zone | cfme-5.9.0.17-TE2  | started        | 14034 | 13507 |      17 | 2018-01-21 |      515 | user_interface:web_services:websocket
 RHV Zone       | cfme-5.9.0.17-TE4  | not responding | 31092 | 13544 |      17 | 2018-01-19 |      237 |
 default        | cfme-5.9.0.17-TE1* | started        | 12783 | 12909 |      17 | 2018-01-21 |      251 | user_interface:web_services:websocket
 default        | cfme-5.9.0.17-TE11 | started        | 13449 | 13622 |      16 | 2018-01-21 |      605 | user_interface:web_services:websocket
 default        | cfme-5.9.0.17-TE12 | started        |  2628 | 14023 |      15 | 2018-01-21 |      532 | user_interface:web_services:websocket

For all rows: Region=0, Version=5.9.0.17, Started=2018-01-18
* marks a master appliance
```

workers
--------

The second table is longer and displays all the workers.

1. the worker name removes some phrases that were common 'ManageIQ', 'Manager', 'Miq', and a number of phrases were trimmed to get the worker name shorter. I could not find a common method for this simplification
2. the port is displayed for workers, but instead of a full url (which all seemed to be of the form `http://127.0.0.1:3000/`) just the protocol and port number have been preserved.
3. We may want to remove pid, spid, and some others, and displaying multiple workers all on a single line.
4. In this environment, the table is 217 lines. It has been truncated in this PR. (since the before does not have zones, hard to get a true before after)

### before

```
 Worker Type                                                              | Status   |     ID |   PID | SPID  | Server id | Queue Name / URL                     | Started On           | Last Heartbeat       | MB Usage
--------------------------------------------------------------------------+----------+--------+-------+-------+-----------+--------------------------------------+----------------------+----------------------+----------
 MiqGenericWorker                                                         | started  | 420334 |  2814 | 14103 |        11 | generic                              | 2018-01-18T23:27:48Z | 2018-01-21T01:30:01Z | 469/500
 MiqGenericWorker                                                         | started  | 420333 |  2807 | 14101 |        11 | generic                              | 2018-01-18T23:27:48Z | 2018-01-21T01:29:59Z | 452/500
 MiqPriorityWorker                                                        | started  | 420336 |  2831 | 14105 |        11 | generic                              | 2018-01-18T23:27:48Z | 2018-01-21T01:30:01Z | 245/600
 MiqPriorityWorker                                                        | started  | 420335 |  2823 | 14104 |        11 | generic                              | 2018-01-18T23:27:48Z | 2018-01-21T01:30:01Z | 273/600
 MiqReportingWorker                                                       | started  | 420344 |  2878 | 14178 |        11 | reporting                            | 2018-01-18T23:28:44Z | 2018-01-21T01:30:00Z | 428/500
 MiqReportingWorker                                                       | started  | 420345 |  2886 | 14181 |        11 | reporting                            | 2018-01-18T23:28:44Z | 2018-01-21T01:30:02Z | 433/500
 MiqScheduleWorker                                                        | started  | 420337 |  2839 | 14107 |        11 |                                      | 2018-01-18T23:27:48Z | 2018-01-21T01:29:57Z | 234/500
 MiqUiWorker                                                              | started  | 420349 |  2918 |       |        11 | http://127.0.0.1:3003                | 2018-01-18T23:28:45Z | 2018-01-21T01:29:51Z | 368/1024
 MiqUiWorker                                                              | started  | 420348 |  2910 |       |        11 | http://127.0.0.1:3002                | 2018-01-18T23:28:45Z | 2018-01-21T01:29:53Z | 399/1024
 ManageIQ::Providers::Vmware::InfraManager::EventCatcher                  | started  | 420247 | 27442 | 13913 |         6 | ems_1                                | 2018-01-18T23:26:40Z | 2018-01-21T01:29:50Z | 287/2048
 ManageIQ::Providers::Vmware::InfraManager::MetricsCollectorWorker        | started  | 420254 | 27450 | 13889 |         6 | vmware                               | 2018-01-18T23:26:40Z | 2018-01-21T01:29:52Z | 454/700
 ManageIQ::Providers::Vmware::InfraManager::MetricsCollectorWorker        | started  | 420255 | 27458 | 13891 |         6 | vmware                               | 2018-01-18T23:26:40Z | 2018-01-21T01:29:52Z | 438/700
 ManageIQ::Providers::Vmware::InfraManager::RefreshWorker                 | started  | 420954 | 20944 | 20769 |         6 | ems_1                                | 2018-01-20T23:36:43Z | 2018-01-21T01:29:55Z | 559/2560
 MiqCockpitWsWorker                                                       | started  | 420257 | 27474 | 13910 |         6 |                                      | 2018-01-18T23:26:40Z | 2018-01-21T01:29:57Z | 263/400
 MiqEmsMetricsProcessorWorker                                             | started  | 420260 | 27482 | 13916 |         6 | ems_metrics_processor                | 2018-01-18T23:26:41Z | 2018-01-21T01:29:58Z | 480/600
 MiqEmsMetricsProcessorWorker                                             | started  | 420262 | 27490 | 13917 |         6 | ems_metrics_processor                | 2018-01-18T23:26:41Z | 2018-01-21T01:29:42Z | 472/600
 ```

### after
```
 Zone           | Type                                      | Status   |   PID | SPID  | Server             | Queue                   | Started    | Heartbeat  | MB Usage
----------------+-------------------------------------------+----------+-------+-------+--------------------+-------------------------+------------+------------+----------
 VMware Zone    | CockpitWs                                 | started  | 27474 | 13910 | cfme-5.9.0.17-TE5  |                         | 2018-01-18 | 2018-01-21 | 263/400
 VMware Zone    | EmsMetricsProcessor                       | started  | 27482 | 13916 | cfme-5.9.0.17-TE5  | ems_metrics_processor   | 2018-01-18 | 2018-01-21 | 480/600
 VMware Zone    | EmsMetricsProcessor                       | started  | 27490 | 13917 | cfme-5.9.0.17-TE5  | ems_metrics_processor   | 2018-01-18 | 2018-01-21 | 472/600
 VMware Zone    | EmsRefreshCore                            | started  | 27505 | 13939 | cfme-5.9.0.17-TE5  | ems_1                   | 2018-01-18 | 2018-01-21 | 319/400
 VMware Zone    | EventHandler                              | started  | 27516 | 13929 | cfme-5.9.0.17-TE5  | ems                     | 2018-01-18 | 2018-01-21 | 329/500
 VMware Zone    | Generic                                   | started  | 27344 | 13680 | cfme-5.9.0.17-TE5  | generic                 | 2018-01-18 | 2018-01-21 | 402/500
 VMware Zone    | Generic                                   | started  | 27352 | 13681 | cfme-5.9.0.17-TE5  | generic                 | 2018-01-18 | 2018-01-21 | 384/500
 VMware Zone    | Priority                                  | started  | 27360 | 13682 | cfme-5.9.0.17-TE5  | generic                 | 2018-01-18 | 2018-01-21 | 376/600
 VMware Zone    | Priority                                  | started  | 27368 | 13683 | cfme-5.9.0.17-TE5  | generic                 | 2018-01-18 | 2018-01-21 | 360/600
 VMware Zone    | Schedule                                  | started  | 27376 | 13684 | cfme-5.9.0.17-TE5  |                         | 2018-01-18 | 2018-01-21 | 237/500
 VMware Zone    | SmartProxy                                | started  | 20659 | 20422 | cfme-5.9.0.17-TE5  | smartproxy              | 2018-01-20 | 2018-01-21 | 538/2048
 VMware Zone    | SmartProxy                                | started  | 20667 | 20424 | cfme-5.9.0.17-TE5  | smartproxy              | 2018-01-20 | 2018-01-21 | 538/2048
 VMware Zone    | Ui                                        | started  | 27554 |       | cfme-5.9.0.17-TE5  | http:3000               | 2018-01-18 | 2018-01-21 | 213/1024
 VMware Zone    | VimBroker                                 | started  | 27574 | 13945 | cfme-5.9.0.17-TE5  | druby:41433             | 2018-01-18 | 2018-01-21 | 403/2048
 VMware Zone    | Vmware::Infra::EventCatcher               | started  | 27442 | 13913 | cfme-5.9.0.17-TE5  | ems_1                   | 2018-01-18 | 2018-01-21 | 287/2048
 VMware Zone    | Vmware::Infra::MetricsCollector           | started  | 27450 | 13889 | cfme-5.9.0.17-TE5  | vmware                  | 2018-01-18 | 2018-01-21 | 454/700
 VMware Zone    | Vmware::Infra::MetricsCollector           | started  | 27458 | 13891 | cfme-5.9.0.17-TE5  | vmware                  | 2018-01-18 | 2018-01-21 | 438/700
 VMware Zone    | Vmware::Infra::Refresh                    | started  | 20944 | 20769 | cfme-5.9.0.17-TE5  | ems_1                   | 2018-01-20 | 2018-01-21 | 559/2560
 VMware Zone    | WebService                                | started  | 27582 |       | cfme-5.9.0.17-TE5  | http:4000               | 2018-01-18 | 2018-01-21 | 214/1024
 VMware Zone    | Websocket                                 | started  | 27609 |       | cfme-5.9.0.17-TE5  | http:5000               | 2018-01-18 | 2018-01-21 | 216/1024
 default        | Generic                                   | started  | 12936 | 12981 | cfme-5.9.0.17-TE1  | generic                 | 2018-01-18 | 2018-01-21 | 506/700
 default        | Generic                                   | started  | 12944 | 12985 | cfme-5.9.0.17-TE1  | generic                 | 2018-01-18 | 2018-01-21 | 532/700

For all rows: Region=0
```

juliancheal commented:
@kbrock this looks neat.

jrafanie commented:
@kbrock there seems to be lots of overlap with `status` and `status_full`.  Is there a use case for having both sets of similar statuses?

jrafanie commented on lib/tasks/evm_application.rb:
```diff
@@ -114,6 +114,149 @@ def self.output_workers_status(servers)
+        "Queue"      => compact_queue(w.queue_name, w.uri),
+        "Started On" => compact_date(w.started_on),
+        "Heartbeat"  => compact_date(w.last_heartbeat),
+        "Ram(mb)"    => compact_memory(w),
```
if we keep both this and status, we should normalize the language.  We have MB Usage in the `status`.  I'm not opposed to changing it but it should probably be the same in both areas.

jrafanie commented on lib/tasks/evm_application.rb:
```diff
@@ -114,6 +114,149 @@ def self.output_workers_status(servers)
+        "Heartbeat"  => compact_date(s.last_heartbeat),
+        "Ram(mb)"    => compact_memory(s),
+        "Roles"      => (s.active_role_names.size == @role_count) ? "*" : s.active_role_names.join(':'),
+      }
```
This feels very similar to the `output_servers_status` method above.  Can we make `rake evm:status` output what you want here and get rid of some of the duplication?  If we really want to keep both, can they both share the same base labels and values but only specify the different labels and values?

jrafanie commented:
> Master?	✔️	✔️	Data is only 1 '*'. Abbreviate title to 'Mstr' or merge with another column?

I can see putting a `*` next to the name of the server for the master server:
cfme-5.9.0.17-TE1*

jrafanie commented:
> Status	✔️	✔️	Simplified to Up/Down/?

I don't know how that buys us much besides losing information.  Status means something to us so capturing the output when things are in a bad state might tell us something and seeing only up or down might lose that info.

jrafanie commented:
> pid	✔️
> spid	✔️

I can see dropping spid but pid is important for both servers and workers

jrafanie commented:
@kbrock I know it's a pain to do on each iteration but can you do a before/after this PR screenshot/embedded text?  It's much easier to grok the changes visually vs. a table.  I think you have only the after text now.

bdunne commented:
![image](https://user-images.githubusercontent.com/651659/37665002-b3a23e8c-2c32-11e8-8efa-a111211bf618.png)
Wow, I totally missed that `*` :smile:

kbrock commented:
thanks @jrafanie updated description to add before/after
there are lots of workers - was too hard to get a true before/after. I could create a gist with the whole before/after if you want. just be prepared to scroll (a lot)

kbrock commented:
>>For all rows: Rgn=0, Version=5.9.0.17, Started=2018-01-18
>@kbrock can we write region? I don't see the value in abbreviating it here

I'll see if I can change the header name in the summary. We already have a special case for region and the value 0

> What if the servers have different versions?
>> cfme-5.9.0.17-TE1*

Anything that is common for all rows shows in the summary. If the values are different, they show in the table itself.

>Can you add a legend at the bottom, where you you log the region, version, started, etc... to say that * indicates the master server?

:+1:

kbrock commented:
I've reviewed the bot's comments and they're rubbish

1. updated Rgn to region
2. added footnote `* marks a master appliance` to help people know what the `* means (thanks LJ / Brandon)
